site/pages/view-all-disruptions.page.tsx

```diff
@@ -94,44 +151,400 @@ export const getWorstSeverity = (severitys: Severity[]): Severity => {
     return worstSeverity;
 };
 
-const ViewAllDisruptions = ({ disruptions, newDisruptionId }: ViewAllDisruptionsProps): ReactElement => {
-    const numberOfDisruptionsPages = Math.ceil(disruptions.length / 10);
-    const [currentPage, setCurrentPage] = useState(1);
+const formatServicesIntoRows = (filter: Filter, setFilter: Dispatch<SetStateAction<Filter>>) => {
+    const cells: { cells: (string | JSX.Element)[] }[] = [];
+
+    filter.services.forEach((service) => {
+        cells.push({
+            cells: [
+                getServiceLabel(service),
+                <button
+                    key={service.id}
+                    className="govuk-link"
+                    onClick={() => {
+                        const { services } = filter;
+                        const indexToRemove = services.findIndex((service) => service.id === service.id);
+                        services.splice(indexToRemove, 1);
+                        setFilter({ ...filter, services });
+                    }}
+                >
+                    Remove
+                </button>,
+            ],
+        });
+    });
+
+    return cells;
+};
+
+const applyDateFilters = (
+    disruptions: TableDisruption[],
+    period: {
+        startTime: string;
+        endTime: string;
+    },
+): TableDisruption[] => {
+    return disruptions.filter((disruption) =>
+        disruption.validityPeriods.some((valPeriod) => {
+            const { startTime, endTime } = valPeriod;
+
+            const periodStartDate = getDate(startTime);
+            const periodEndDate = endTime ? getDate(endTime) : undefined;
+
+            return filterDatePeriodMatchesDisruptionDatePeriod(
+                getFormattedDate(period.startTime),
+                getFormattedDate(period.endTime),
+                periodStartDate,
+                periodEndDate,
+            );
+        }),
+    );
+};
+
+export const filterDisruptions = (disruptions: TableDisruption[], filter: Filter): TableDisruption[] => {
```

Comment:
I wonder if we could streamline this function by doing something like:

```ts
export const filterDisruptions = (disruptions: TableDisruption[], filter: Filter): TableDisruption[] => {
    let disruptionsToDisplay = disruptions;

    disruptionsToDisplay = disruptionsToDisplay.filter((disruption) => {
        if (filter.services.length > 0) {
            const filterServiceRefs = filter.services.map((service) => service.id.toString());
            const disruptionServiceRefs = disruption.serviceIds;
            let showService = false;

            disruptionServiceRefs.forEach((disruptionServiceRef) => {
                if (filterServiceRefs.includes(disruptionServiceRef)) {
                    showService = true;
                }
            });

            if (!showService) {
                return false;
            }
        }

        if (filter.mode) {
            const swappedMode = getDisplayByValue(VEHICLE_MODES, filter.mode);

            if (!swappedMode || !disruption.modes.includes(swappedMode)) {
                return false;
            }
        }

        if (filter.severity && disruption.severity !== filter.severity) {
            return false;
        }

        if (filter.status && disruption.status !== filter.status) {
            return false;
        }

        if (filter.operators.length > 0) {
            const filterOperatorsRefs = filter.operators.map((op) => op.operatorRef);

            if (!disruption.operators.some((operator) => filterOperatorsRefs.includes(operator.operatorRef))) {
                return false;
            }
        }

        return true;
    });

    if (filter.period) {
        disruptionsToDisplay = applyDateFilters(disruptionsToDisplay, filter.period);
    }

    return disruptionsToDisplay;
};
```

This way we'd only be looping over the disruptions a couple of times

Reply:
Added.